File: 103.binary-tree-zigzag-level-order-traversal.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func zigzagLevelOrder(root *TreeNode) [][]int {
	if root == nil {
		return nil
	}
	q := []*TreeNode{root}
	res := [][]int{}
	index := 0
	for len(q) > 0 {
		l := len(q)
		row := []int{}
		if index == 0 {
			for i := 0; i < l; i++ {
				row = append(row, q[i].Val)
			}
		} else {
			for i := l - 1; i >= 0; i-- {
				row = append(row, q[i].Val)
			}
		}
		res = append(res, row)
		index = (index + 1) % 2
		for i := 0; i < l; i++ {
			n := q[0]
			q = q[1:]
			if n.Left != nil {
				q = append(q, n.Left)
			}
			if n.Right != nil {
				q = append(q, n.Right)
			}
		}
	}
	return res
}

File: 103.binary-tree-zigzag-level-order-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        List<List<Integer>> res = new ArrayList<>();
        if (root == null) {
            return res;
        }
        Queue<TreeNode> q = new LinkedList<>();
        q.add(root);
        boolean reverse = false;
        while (!q.isEmpty()) {
            List<Integer> list = new ArrayList<>();
            int l = q.size();
            for (int i = 0; i < l; i++) {
                TreeNode n = q.poll();
                if (reverse) {
                    list.add(0, n.val);
                } else {
                    list.add(n.val);
                }
                if (n.left != null) {
                    q.add(n.left);
                }
                if (n.right != null) {
                    q.add(n.right);
                }
            }
            res.add(list);
            reverse = !reverse;
        }
        return res;
    }
}
